packages/dds/tree/src/shared-tree/treeAlpha.ts

```diff
@@ -487,8 +546,123 @@ export const TreeAlpha: TreeAlpha = {
 		const parentSchema = treeNodeApi.schema(parent);
 		return getPropertyKeyFromStoredKey(parentSchema, storedKey);
 	},
+
+	child: (
+		node: TreeNode,
+		propertyKey: string | number,
+	): TreeNode | TreeLeafValue | undefined => {
+		const flexNode = getOrCreateInnerNode(node);
+		debugAssert(
+			() => !flexNode.context.isDisposed() || "The provided tree node has been disposed.",
+		);
+
+		const schema = treeNodeApi.schema(node);
+
+		switch (schema.kind) {
+			case NodeKind.Array: {
+				const sequence = flexNode.tryGetField(EmptyKey) as FlexTreeSequenceField | undefined;
+
+				// Empty sequence - cannot have children.
+				if (sequence === undefined) {
+					return undefined;
+				}
+
+				const index =
+					typeof propertyKey === "number"
+						? propertyKey
+						: asIndex(propertyKey, Number.POSITIVE_INFINITY);
+
+				// If the key is not a valid index, then there is no corresponding child.
+				if (index === undefined) {
+					return undefined;
+				}
```

Comment:
It's not clear to me that we should be tolerating strings containing numbers to be used as keys into Arays here.

From reading our APi docs and types, I would not expect this to be supported.

```suggestion
				// If the key is not a valid index, then there is no corresponding child.
				if (typeof propertyKey !== "number") {
					return undefined;
				}
```

Reply:
Actually, I don't think that's what we want. POJO arrays allow both, so we probably should as well.
https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAhgJwXAnjAvDA2gRgDQBMeAzALoDcAUJaJCADYCmAdPSAOYAUiyKWADKQCU5GAHoxMHDXAQGLNlx6osAIn6rhoiVOq05TVh04AlRgDMmwKMwAWcCNySo8MfkJHjJUBAFdGMnSGiqYWVjb2jsooruqqHtrefoxAA

Reply:
Similarly, Array nodes allow string keys.
```typescript
class TestArray extends schema.array("TestObject", schemaFactory.string) {}
const test = new TestObject(["Hello world!"]);

const zero = test["0"]; // "Hello world!"
```

Reply:
It is literally impossible to implement arrays without supporting string keys, since JavaScript does not support numeric keys: they are stringified. The property key received by a proxy can never be a number: it is always a string, even when the user provides a number.

Our arrays support string keys exactly the same as JavaScript arrays, meaning we use typescript to tell you to pass in a number, which JS converts to a string behind the scenes. We do not support strings in places where not supporting strings is possible, for example in "at".

Our code for converting string indexes back to numbers exists to make indexing with numbers work: the fact indexing with strings works is an unfortunate side effect of this very questionable language.

Reply:
For example ["x"]["0"] gives "x" just like ["x"][0] does, and the proxy can't tell those two lookups apart as both search for a member under the key "0". Note that this also gives "x": `{"0": "x"}[0]` : this conversion from number to string is a JS property thing, and has nothing to do with arrays.

Reply:
I think my main point can be summed up by the following:

Array nodes can be indexed using string-formatted integer keys, so I don't think it makes sense for them not to work via `child`.

```typescript
class MyArray extends schemaFactory.array("my-array", schemaFactory.string) {}
const myArray = new MyArray(["Hello world"]);
myArray["0"]; // "Hello world"
Tree.child(myArray, "0"); // Should match
```

- Similarly, for object nodes, we support indexing them with numeric keys, so the behavior of `child` should match.

```typescript
class MyObject extends schema.object("MyObject", {
	"1": SchemaFactory.optional(schema.string),
}) {}
const myObject = new MyObject({ 1: "Hello world" });

myObject[1]; // "Hello world"
TreeAlpha.child(myObject, 1); // Should match
```

Maps in TS with string keys _do not_ support numeric indexing, so not implicitly converting integers to strings in that context makes more sense. An argument could be made for supporting them for consistency with the behavior of `child` for the other node kinds, but I don't think this is nearly as important.

Reply:
FWIW, @Josmithr and I talked offline and I think the idea of consistency with POJO is compelling. I strongly disagree with the idea of possibly wanting maps to support something similar, as I think a very important property of maps is that keys that compare equal should look up the same value, and keys that do not should not. If we break that contract, I think this would be a massive bug pit, and I don't see any offsetting reason to do so. Setting that aside, I'll vote in favor of consistency with POJO :)